### PR TITLE
NN-5054 NN-5057 sanction integration and removal of punishments

### PR DESCRIPTION
**Pull request by timoj58**
integration of nomis sanctions, and refactoring caution and damages owed into punishments.  Note front end will still be presented the data as part of the outcome DTO to avoid extensive re-write of front end.

This PR is currently only draft as can not be merged until prison api is available to be called

### CODE REVIEW

**Review by neilmendum — COMMENTED** [review]
Appreciate this is still in draft. Added a few comments about tests and function readability

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/reported/PunishmentsService.kt [hunk 1]
```diff
@@ -115,6 +158,93 @@ class PunishmentsService(
...
+      punishment.sanctionSeq = prisonApiGateway.createSanction(
+        adjudicationNumber = reportedAdjudication.reportNumber,
+        sanction = punishment.mapPunishmentToSanction(),
+      )
+    }
+  }
+
+  private fun handleCautionChange(reportedAdjudication: ReportedAdjudication, caution: Boolean) {
+    when (val cautionPunishment = reportedAdjudication.punishments.firstOrNull { it.type == PunishmentType.CAUTION }) {
+      null -> if (caution) amendToCaution(reportedAdjudication = reportedAdjudication)
```
[the remark this answers is not in the document]
Comment (timoj58) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/reported/PunishmentsService.kt:204]: yeah the unit tests are in the punishmentsServiceTests with paramterization.  Its not visible until you expand that file as github has collapsed it

[the remark this answers is not in the document]
Comment (timoj58) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/reported/PunishmentsService.kt:204]: jacoco report as 100% line and branch for this function too.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/reported/PunishmentsService.kt [hunk 2]
```diff
@@ -115,6 +158,93 @@ class PunishmentsService(
     }.flatten()
   }
 
+  private fun handleDamagesOwedChange(reportedAdjudication: ReportedAdjudication, amount: Double?) {
```
[the remark this answers is not in the document]
Comment (timoj58) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/reported/PunishmentsService.kt:161]: ditto this also has tests

[the remark this answers is not in the document]
Comment (timoj58) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/reported/PunishmentsService.kt:161]: this has tests as well, but is already merged hence why its not in the PR.  see 
 inner class GetSuspendedPunishments {

[the remark this answers is not in the document]
Comment (timoj58) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/reported/PunishmentsService.kt:161]: This has tests, they are in the file punishmentsServiceTests you need to expans.  100% line and branch coverage according to jacoco

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/reported/PunishmentsService.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (timoj58) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/reported/PunishmentsService.kt:204]: the reason behind this is caution is the value passed from the UI (either true of false). 
Caution punishment is the record we may have saved previously, if caution was true.

Therefore, based on what we receive (true or false) and what we already persist (caution, or nothing) need to consider next steps.  as per logic.

So if we have no caution punishment on file, and we are now a caution, we need to remove all the other sanctions in nomis (and our system) and create a caution.

If we are no longer a caution, we simply remove the record from nomis (and our own system).

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/reported/PunishmentsService.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (timoj58) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/reported/PunishmentsService.kt:161]: Screenshot from 2023-04-27 12-40-26

**Review by neilmendum — APPROVED** [review]

**Pull request merged**